Features/logging

MaFeLP opened the pull request:
This adds support for log files and makes the console much nicer looking!

## Features
* Logs log entries to:
  * `STDOUT` For log levels (`TRACE` to[^1]) `INFO`
  * `STDERR` For log levels `WARN` and `ERROR`
  * logfile: For all log levels[^1]
* Highly configurable

## Configuration
### File
The logger can be configured via a `log4rs.yml` file ([documentation](https://docs.rs/log4rs/latest/log4rs/index.html#configuration-via-a-yaml-file)). **THIS IS NOT RECOMMENDED AND REMOVES SOME FUNCTIONALITY[^2]**

## Environment Variables
| Variable | Description | Default |
| --- | --- | --- |
| `LOGGING_PATTERN` | Describes how log entries should be shown/logged (see [documentation](https://docs.rs/log4rs/latest/log4rs/encode/pattern/index.html) | `{h({d(%Y-%m-%d %H:%M:%S)} [{t}/{l}]: {m:>10}{n})}` |
| `LOGGING_ROLLOVER_SIZE` | At which file size the log file should be rolled over[^3] | `10000000` (10MB) |
| `LOGGING_FOLDER` | The folder in which all logs should be saved | `logs` |
| `LOGGING_ARCHIVE_PATTERN` | the pattern with which log archives are being save. Must include at least one `{}` to represent an index. (See [documentation](https://docs.rs/log4rs/latest/log4rs/append/rolling_file/policy/compound/roll/fixed_window/struct.FixedWindowRollerDeserializer.html#configuration) | `{}.log.gz` |
| `LOGGING_FILE_COUNT` | How many old log files should be stored. | `10` |
| `LOGGING_LEVEL_THRESHOLD` | The minimum log level required for entries to pass | `INFO` |


## New Dependencies
* [log](https://crates.io/crates/log): For the logging macros
* [log4rs](https://crates.io/crates/log4rs): Logging framework for printing log entries to the console and the file
* [anyhow](https://crates.io/crates/anyhow): To provide the correct return type for custom log4rs modules

[^1]: By default, development builds log entries starting `DEBUG`. On release builds, log entries start at level `INFO`.
[^2]: Log files are not rolled over[^3] after a restart and there is no way to log INFO to stdout and warn to stderr without also logging warn to stdout!
[^3]: The file handle to the current file is closed, the file is archived as (by default) `0.log.gz` and all are log files are counted up by one.

MaFeLP committed:
Added log and log4rs dependencies

Signed-off-by: MaFeLP <60669873+MaFeLP@users.noreply.github.com>

MaFeLP committed:
Added logger configuration

Signed-off-by: MaFeLP <60669873+MaFeLP@users.noreply.github.com>

MaFeLP committed:
Added logs folder to the gitignore

Signed-off-by: MaFeLP <60669873+MaFeLP@users.noreply.github.com>

MaFeLP committed:
Fixed compiling issue

Signed-off-by: MaFeLP <60669873+MaFeLP@users.noreply.github.com>

MaFeLP committed:
Fixed double output in log file.

Signed-off-by: MaFeLP <60669873+MaFeLP@users.noreply.github.com>

MaFeLP committed:
Fixed warn being displayed twice.

Signed-off-by: MaFeLP <60669873+MaFeLP@users.noreply.github.com>

MaFeLP committed:
Made the logging pattern universal

Signed-off-by: MaFeLP <60669873+MaFeLP@users.noreply.github.com>

MaFeLP committed:
Updated the log format to look nicer

Signed-off-by: MaFeLP <60669873+MaFeLP@users.noreply.github.com>

MaFeLP committed:
📗 Added documentation to logger_init.rs

Signed-off-by: MaFeLP <60669873+MaFeLP@users.noreply.github.com>

MaFeLP committed:
Added some TODOs

Signed-off-by: MaFeLP <60669873+MaFeLP@users.noreply.github.com>

MaFeLP committed:
Fixed older logs being put into the wrong directory

Signed-off-by: MaFeLP <60669873+MaFeLP@users.noreply.github.com>

MaFeLP committed:
Made the logger more configurabel via environment variables

Signed-off-by: MaFeLP <60669873+MaFeLP@users.noreply.github.com>

MaFeLP committed:
Also made warnings be printed out

Signed-off-by: MaFeLP <60669873+MaFeLP@users.noreply.github.com>

MaFeLP committed:
Added anyhow dependency

Signed-off-by: MaFeLP <60669873+MaFeLP@users.noreply.github.com>

MaFeLP committed:
Added custom rollover trigger

This is the same as the log4rs SizeTrigger, but it also
rolls over the file, if an old file exists, before the first log entry.

Signed-off-by: MaFeLP <60669873+MaFeLP@users.noreply.github.com>

MaFeLP committed:
Outsourced logger_init to its own module and split up components

Signed-off-by: MaFeLP <60669873+MaFeLP@users.noreply.github.com>

MaFeLP committed:
Reformatted the logger module with rustfmt

Signed-off-by: MaFeLP <60669873+MaFeLP@users.noreply.github.com>

MaFeLP committed:
Added documentation

Signed-off-by: MaFeLP <60669873+MaFeLP@users.noreply.github.com>

MaFeLP committed:
Added custom build script to reference the git ref as environment.

Signed-off-by: MaFeLP <60669873+MaFeLP@users.noreply.github.com>

MaFeLP committed:
Changed print statements to log statements in main.rs

Signed-off-by: MaFeLP <60669873+MaFeLP@users.noreply.github.com>

MaFeLP committed:
Changed print statements to log statements in config.rs

Signed-off-by: MaFeLP <60669873+MaFeLP@users.noreply.github.com>

MaFeLP committed:
Fixed too long messages being cut off.

Signed-off-by: MaFeLP <60669873+MaFeLP@users.noreply.github.com>

MaFeLP committed:
Fixed missing bracket at the end of the git hash

Signed-off-by: MaFeLP <60669873+MaFeLP@users.noreply.github.com>

MaFeLP committed:
Added logging to the replies module

Signed-off-by: MaFeLP <60669873+MaFeLP@users.noreply.github.com>

MaFeLP committed:
Added logging to the kaenguru module

Signed-off-by: MaFeLP <60669873+MaFeLP@users.noreply.github.com>

MaFeLP committed:
Moved print statements to log recods in xd.rs

Signed-off-by: MaFeLP <60669873+MaFeLP@users.noreply.github.com>

MaFeLP committed:
Bump version to 0.3.0

Signed-off-by: MaFeLP <60669873+MaFeLP@users.noreply.github.com>

G0rger commented on src/logger/custom/trigger.rs:
```diff
@@ -0,0 +1,70 @@
+    /// It is set, by [default_logger](crate::logger::default_logger) before the configuration
+    /// is constructed. It **should only be unset by [CustomTrigger::trigger]**, when the logfile
+    /// has been rolled over **ONCE** at program startup!
+    pub static ref LOG_FILE_EXISTS: Arc<Mutex<bool>> = Arc::new(Mutex::new(false));
```
don't make this pub if it should only be used in this module

G0rger replied:
```suggestion
    pub static ref LOG_FILE_EXISTS: Mutex<bool> = Mutex::new(false);
```

MaFeLP committed:
Added the ability to use a custom logger configuration.

Signed-off-by: MaFeLP <60669873+MaFeLP@users.noreply.github.com>

MaFeLP committed:
Fixed compiling issues

Signed-off-by: MaFeLP <60669873+MaFeLP@users.noreply.github.com>

MaFeLP committed:
Removed unneeded log4rs components

Signed-off-by: MaFeLP <60669873+MaFeLP@users.noreply.github.com>

MaFeLP committed:
Added endline at the end of the file

Signed-off-by: MaFeLP <60669873+MaFeLP@users.noreply.github.com>

G0rger commented on src/logger/custom/filter.rs:
```diff
@@ -0,0 +1,56 @@
+}
+
+impl UpperThresholdFilter {
+    /// Creates a new `ThresholdFilter` with the specified maximum logging
```
```suggestion
    /// Creates a new `UpperThresholdFilter` with the specified maximum logging
```

G0rger commented on src/logger/mod.rs:
```diff
@@ -0,0 +1,289 @@
+            Box::new(
+                FixedWindowRoller::builder()
+                    .build(
+                        format!("{}/{}", folder, archive_pattern).as_str(),
```
```suggestion
                        &format!("{}/{}", folder, archive_pattern),
```

G0rger commented on src/logger/mod.rs:
```diff
@@ -0,0 +1,289 @@
+    log4rs::init_config(config).unwrap();
+
+    for msg in warnings {
+        warn!(target: "xdbot", "{}", msg);
```
```suggestion
        warn!(target: "xd_bot", "{}", msg);
```

MaFeLP committed:
Fixed typos

Co-authored-by: Jannis Fricke <jannis@gorger.dev>

MaFeLP committed:
Remove unneeded atomic reference counting

As `Arc<_>` uses atomic reference counting, its access times are much higher. Through the lazy_static there always is a reference, which does not need to be dropped, when nothing is accessing it.

Co-authored-by: Jannis Fricke <jannis@gorger.dev>

G0rger commented on src/logger/custom/filter.rs:
```diff
@@ -0,0 +1,56 @@
+//!
+//! This module contains all custom logging filters. A filter examines a record which is to be
+//! logged and then  decides, if it should be logged or not.
```
```suggestion
//! logged and then decides, if it should be logged or not.
```

G0rger commented on src/logger/custom/trigger.rs:
```diff
@@ -0,0 +1,70 @@
+    /// It is set, by [default_logger](crate::logger::default_logger) before the configuration
+    /// is constructed. It **should only be unset by [CustomTrigger::trigger]**, when the logfile
+    /// has been rolled over **ONCE** at program startup!
+    pub static ref LOG_FILE_EXISTS: Mutex<bool> = Mutex::new(false);
```
I just noticed, why not use an [AtomicBool](https://doc.rust-lang.org/std/sync/atomic/struct.AtomicBool.html)? It should be faster than a Mutex with bool.

MaFeLP replied:
See 03c7b33

MaFeLP replied:
> Why did you revert to `Mutex<bool>`?

The first way in the documentation didn't compile, so I changed it in 0e8519fd4b3afbfeeb172282fe43d4fa3b465227

MaFeLP committed:
Fixed typo

Co-authored-by: Jannis Fricke <jannis@gorger.dev>

G0rger commented on src/logger/custom/trigger.rs:
```diff
@@ -0,0 +1,71 @@
+//! * [CustomTrigger]
+//!
+
+use log4rs::append::rolling_file::{policy::compound::trigger::Trigger, LogFile};
+use std::sync::{Arc, Mutex};
+use std::sync::atomic::AtomicBool;
```
```suggestion
use log4rs::append::rolling_file::{policy::compound::trigger::Trigger, LogFile};
use std::sync::atomic::AtomicBool;
```

G0rger commented:
Why did you revert to `Mutex<bool>`?

MaFeLP committed:
Made LOG_FILE_EXISTS an atomic bool.

This is on @G0rger's Suggestion (https://github.com/MaFeLP/discord_bots/pull/12#discussion_r777656053)

Signed-off-by: MaFeLP <60669873+MaFeLP@users.noreply.github.com>

MaFeLP commented:
> Why did you revert to `Mutex<bool>`?

The first way in the documentation didn't compile, so I changed it in 0e8519fd4b3afbfeeb172282fe43d4fa3b465227

MaFeLP committed:
Removed unused imports

Signed-off-by: MaFeLP <60669873+MaFeLP@users.noreply.github.com>

G0rger commented:
> > Why did you revert to `Mutex<bool>`?
> 
> The first way in the documentation didn't compile, so I changed it in 0e8519fd4b3afbfeeb172282fe43d4fa3b465227

I think you forgot the `.get_mut()` in the AtomicBool.

MaFeLP commented:
> > > Why did you revert to `Mutex<bool>`?
> > 
> > 
> > The first way in the documentation didn't compile, so I changed it in [0e8519f](https://github.com/MaFeLP/discord_bots/commit/0e8519fd4b3afbfeeb172282fe43d4fa3b465227)
> 
> I think you forgot the `.get_mut()` in the AtomicBool.

No, I did not.

I tried multiple approaches for this, but cargo said that `LOGGING_FILE_EXISTS.get_mut()` can not be dereferenced:

```log
error[E0596]: cannot borrow data in a dereference of `*LOG_FILE_EXISTS` as mutable
  --> src/logger/custom/trigger.rs:62:25
   |
62 |         let mut test = *LOG_FILE_EXISTS.get_mut();
   |                         ^^^^^^^^^^^^^^^ cannot borrow as mutable
   |
   = help: trait `DerefMut` is required to modify through a dereference, but it is not implemented for `LOG_FILE_EXISTS`
```

G0rger commented on src/logger/custom/trigger.rs:
```diff
@@ -0,0 +1,70 @@
+    ///
+    /// returns: Result<bool, Error>
+    fn trigger(&self, logfile: &LogFile) -> anyhow::Result<bool> {
+        let log_file_exists = LOG_FILE_EXISTS.load(Ordering::Relaxed);
+        let roll: bool = (logfile.len_estimate() > self.limit) || log_file_exists;
+
+        if log_file_exists {
+            LOG_FILE_EXISTS.store(false, Ordering::Relaxed);
+        }
+
+        Ok(roll)
```
I think it would more appropriate if [`compare_exchange`](https://doc.rust-lang.org/std/sync/atomic/struct.AtomicBool.html#method.compare_exchange) would be used:
```suggestion
        let log_file_exists = LOG_FILE_EXISTS.compare_exchange(true, false, Ordering::Relaxed, Ordering::Relaxed).is_ok();
        let roll: bool = (logfile.len_estimate() > self.limit) || log_file_exists;

        Ok(roll)
```

MaFeLP replied:
using `compare_exchange` makes the code only more cluttered and harder to understand, as code like **this** would be required, because `compare_exchange` returns a result which must first be interpreted:

```rust
        let log_file_roll = match LOG_FILE_EXISTS.compare_exchange(
            true,
            false,
            Ordering::Relaxed,
            Ordering::Relaxed,
        ) {
            Ok(_) => true,
            Err(_) => false,
        };
        let roll: bool = (logfile.len_estimate() > self.limit) || log_file_roll;

        Ok(roll)
```

G0rger replied:
But without it there's a race condition with a possibility of two log rollovers by two simultaneous log entries.

G0rger replied:
Plus with a call to `is_ok` on the `Result<bool, bool>` does the match for you.

MaFeLP replied:
No, because the first log calls in `main.rs` (or `logger/mod.rs`) are **not** asynchronous, which means that only one log record is logged at a time.

Afterwards the log calls only have `LOG_FILE_EXISTS = false`, which could only trigger a size roll over.

G0rger replied:
Still

MaFeLP replied:
Is still not that easy to understand

G0rger reviewed (dismissed):
LGTM

MaFeLP committed:
Fixed bug

latest.log would not show the first log entry, as it would be written
to 0.log.gz instead, if another run was conducted before.

Signed-off-by: MaFeLP <60669873+MaFeLP@users.noreply.github.com>

G0rger reviewed (approved):
LGTM
